### PR TITLE
feat(controller): Basic support for multi-cluster workflows. Closes #3523

### LINKED ISSUE #3523: Multi-cluster, multi-namespace workflows
# Summary

Run workflows across multiple clusters.

# Motivation

So you only need to run one Argo Workflows installation.
So you can run a workflow that has nodes in different clusters.

# Proposal

Like Argo CD.

#3516

---

**Message from the maintainers**:

If you wish to see this enhancement implemented please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by alexec**
Closes #3523

Signed-off-by: Alex Collins <alex_collins@intuit.com>

Checklist:

* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

### CODE REVIEW

File: cmd/argo/commands/cluster/rm.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/argoproj/argo/cmd/argo/commands/client"
+)
+
+func NewRMCommand() *cobra.Command {
+	return &cobra.Command{
+		Use: "rm CLUSTER_NAME",
```
Comment (simster7) [inline cmd/argo/commands/cluster/rm.go:19]: `delete`? Consistent with `argo {template,cron} delete`

Comment (simster7) [reply cmd/argo/commands/cluster/rm.go:19]: Perhaps not since it doesn't delete the cluster, but removes it from being tracked

Comment (alexec) [reply cmd/argo/commands/cluster/rm.go:19]: good point - I copied from Argo CD - but @jessesuen explained a reason for this to me - I think it was that you were not actually deleting the cluster, just removing it from from the config - but I feel not very strongly at all about this

**Comment by alexec** [comment]
You've some codegen changes. My recommendation is to copy the patch from the lgos and use `pbpaste|wc -c 30-|git apply` to apply it. Alternatively, delete the changed files and run `make codegen`.

**Comment by alexec** [comment]
Closing PR so it does not distract me.

**Comment by alexec** [comment]
*  TestArchiveStrategies
* TestArtifactRepositoryRef
* TestDefaultParameterOutputs
* TestGlobalScope
* TestSameInputOutputPathOptionalArtifact
* TestPermission
* TestMetaDataNamespace

File: workflow/controller/exec_control.go [hunk 2]
```diff
@@ -20,6 +20,10 @@ func (woc *wfOperationCtx) applyExecutionControl(pod *apiv1.Pod, wfNodesLock *sy
 	if pod == nil {
 		return nil
 	}
+	clusterName, ok := pod.Labels[common.LabelKeyClusterName]
```
Comment (alexec) [inline workflow/controller/exec_control.go:23]: fix

File: workflow/controller/indexes/workflow_index.go [hunk 3]
```diff
@@ -17,11 +17,15 @@ func MetaWorkflowIndexFunc(obj interface{}) ([]string, error) {
 	if err != nil {
 		return []string{}, fmt.Errorf("object has no meta: %v", err)
 	}
+	namespace, ok := m.GetLabels()[common.LabelKeyWorkflowNamespace]
```
Comment (alexec) [inline workflow/controller/indexes/workflow_index.go:20]: fix

File: workflow/controller/operator.go [hunk 4]
```diff
@@ -344,13 +346,12 @@ func (woc *wfOperationCtx) operate() {
...
-		// the error are handled in the callee so just log it.
-		msg := "error in entry template execution"
-		woc.log.WithError(err).Error(msg)
-		msg = fmt.Sprintf("%s %s: %+v", woc.wf.Name, msg, err)
 		switch err {
 		case ErrDeadlineExceeded:
-			woc.eventRecorder.Event(woc.wf, apiv1.EventTypeWarning, "WorkflowTimedOut", msg)
+			woc.eventRecorder.Event(woc.wf, apiv1.EventTypeWarning, "WorkflowTimedOut", fmt.Sprintf("%s error in entry template execution: %+v", woc.wf.Name, err))
+		}
+		if !woc.wf.Status.Phase.Completed() {
```
Comment (alexec) [inline workflow/controller/operator.go:353]: pork-barrel

**Pull request closed without merging**